### ISSUE TITLE
fix(transformer): copy FFN output when PLE is configured but per-layer input is absent

### ISSUE THREAD

**Issue by geisten** [geisten/geistlib#326]
Severity: HIGH

## What happened

`src/archs/transformer/forward/layer_ffn.c:326` writes the FFN result to `scratch_h_post_ff` when PLE is enabled. `src/archs/transformer/forward/layer_ple.c:55-176` only materializes `h_out` when `apply_ple` is true and a non-null per-layer input is supplied. Otherwise it returns `GEIST_OK` without copying the FFN result.

`transformer_forward_one_layer` currently accepts a null per-layer input, so callers can receive stale/zero output while the API reports success.

Observed reproduction:

```sh
GEIST_GGUF_PATH=gguf_artifacts/gemma4-e2b-Q4_K_M.gguf   bin/mac-omp/release/tests/test_state_layer_fwd_int
```

Layers 0, 4, and 15 produced zero output in the isolated layer-forward test (and layer 19 with Q4_K predecode disabled).

## Expected

Either:

1. copy `scratch_h_post_ff` to `h_out` when PLE is skipped, or
2. reject the invalid PLE/input combination before executing the layer.

The API contract should make the chosen behavior explicit.

## Acceptance criteria

- Isolated layer-forward tests never return success with untouched output.
- Regression coverage includes PLE layers with present and absent per-layer input.
- Full forward parity remains within the existing tolerance.

## Mandatory C23 implementation rules

- In every modified array API, place all length/capacity parameters before the arrays they describe; use VLA contracts such as `T data[static len]` where the non-null/minimum-size precondition is real.
- Use C23 `bool`, `true`, `false`, and `nullptr`/`nullptr_t` instead of integer booleans or `NULL`. Replace constant macros with typed `constexpr` objects when compile-time and ABI requirements permit.
- Use `[[nodiscard]]` for returned status/size values that must be checked. Add `restrict` only when non-aliasing is proven for every caller. Use `typeof`/`auto` only when they improve clarity without hiding ownership or changing code generation.
- Perform checked `size_t` arithmetic before allocation, indexing, alignment, or pointer advancement. For cursors, validate against `(size_t)(end - p)` before advancing; do not use `p + len` or `&p[len]` as the bounds check.
- Preserve public ABI unless the ticket explicitly includes a versioned compatibility migration.

## Mandatory performance and quality gates

- Establish an unchanged-parent baseline, then run the layer-forward tests and relevant benchmark after every change to FFN/PLE copying or routing.
- Run the macOS prefill/decode sweep at representative short and long contexts (including 128, 512, and 1024) with identical model hash, flags, threads, and environment.
- Existing full-forward logits/known-answer quality must remain unchanged within existing tolerances; do not relax them.
- No repeatable tok/s, latency, or memory regression is acceptable. Store raw before/after results in the PR.

**Comment by geisten** [comment]
Fixed in 6674a5b (PR #344), released in **v0.10.2**.

**The report was right, and the diagnosis needs one correction.** The `else` branch this ticket describes was not missing by oversight — 81f5a98 ("perf(metal): keep qwen35 attention on device") deleted it deliberately, having made the FFN write its residual result straight into `h_out_buf` when PLE is off. That is a valid optimisation. What it missed is that the two stages then branched on *different* conditions:

- `layer_ffn.c` picked its destination on `apply_ple`
- `layer_ple.c` ran on `apply_ple && per_layer_input_buf != nullptr`

The combination in between is the bug: a PLE family called with no per-layer input routes the FFN result into `scratch_h_post_ff`, then skips the stage that would have read it, and `h_out` is never written at all.

**Fix: stop having two conditions.** `ctx->run_ple` is computed once in `transformer_layer_ctx_init` and both stages read it, so the destination and the execution cannot disagree again.

**Scope.** Prefill and decode always pass a per-layer input for a PLE model, so the two conditions agree there and **shipped inference was unaffected**. What broke is `transformer_forward_one_layer` and any caller taking the documented "NULL to skip the PLE block" at its word.

**Acceptance criteria.**
- [x] Isolated layer-forward tests never return success with untouched output — `run_layer_ple` poisons `h_out` with a sentinel before every call and checks every element after
- [x] Regression coverage for PLE layers with present *and* absent per-layer input — absent must equal present-and-zero bit for bit, and present-and-non-zero must differ from it, so the equivalence cannot hold vacuously (777 of 1536 outputs move)
- [x] Full forward parity within existing tolerance — known-answer, decode, determinism, chat and multi-session all pass

**One process note.** My first pass on this ticket concluded the defect was not present; that assessment was made against a branch base ~40 commits behind main, which predated 81f5a98. The equivalence test added in 7b0578e is what caught it after rebasing — it reported all 1536 outputs disagreeing. Both commits are in the PR.

**Performance gate.** Best-of-5 totals at 128/512/1024 against `origin/main` on the same machine: 1490/4119/7840 ms vs 1509/4133/7808; RSS 4470/4480/4493 MB vs 4467/4478/4491.
